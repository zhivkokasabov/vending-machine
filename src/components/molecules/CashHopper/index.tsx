import React from "react";

import { Coin } from "components/atoms/Coin";
import { useCashHopper } from "providers/CashHopper.provider";

export const CashHopper = () => {
  const { stashedCoins } = useCashHopper();

  const stackDefaultValue: Record<string, number> = {};
  const groupedCoins = stashedCoins.reduce((acc, curr) => {
    const roundedKey = curr.toFixed(2);

    acc[roundedKey] = acc[roundedKey] ? acc[roundedKey] + 1 : 1;

    return acc;
  }, stackDefaultValue);
  // might sort them eventually, we'll see
  const stacks = Object.entries(groupedCoins);

  return (
    <div className="flex gap-2 pt-4 w-full max-w-full flex-wrap" data-testid="cash-hopper">
      {stacks.map(([denomination, stackSize]) => {
        return <div key={denomination} className="relative w-[3rem] h-[3rem]">
          {
            Array.from({ length: stackSize > 5 ? 5 : stackSize }, (_, index) => {
              return <Coin
                key={index}
                denomination={denomination}
                // generate for: top-[-3px] top-[-6px] top-[-9px] top-[-12px] top-[-15px]
                className={`absolute left-0 top-[${index * -3}px]`}
              />;
            })
          }
          {
            stackSize < 5
              ? (null)
              : (
                <div className="absolute bottom-[0.5rem] right-[-1px] rounded-full bg-white border border-black text-center w-4 h-4 text-xs">{stackSize}</div>
              )
          }
        </div>
      })}
    </div>
  )
}