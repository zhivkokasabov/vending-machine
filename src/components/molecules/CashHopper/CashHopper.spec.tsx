import { CashHopper } from ".";

import { render } from "@testing-library/react";

import * as cashHopper from "providers/CashHopper.provider";

jest.mock("providers/CashHopper.provider");

describe("CashHopper tests", () => {

  beforeEach(() => {
    jest.spyOn(cashHopper, 'useCashHopper').mockReturnValue({
      stashCoins: jest.fn(),
      stashedCoins: []
    });
  });

  const renderComponent = () => render(
    <CashHopper />
  );

  it("should render", () => {
    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should group coins", () => {
    jest.spyOn(cashHopper, 'useCashHopper').mockReturnValue({
      stashCoins: jest.fn(),
      stashedCoins: [1, 2, 2, 1]
    });

    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });

  it("should not stack more than 5 coins", () => {
    jest.spyOn(cashHopper, 'useCashHopper').mockReturnValue({
      stashCoins: jest.fn(),
      stashedCoins: [1, 2, 2, 1, 2, 2, 2, 2, 2]
    });

    const { asFragment } = renderComponent();

    expect(asFragment()).toMatchSnapshot();
  });
});
