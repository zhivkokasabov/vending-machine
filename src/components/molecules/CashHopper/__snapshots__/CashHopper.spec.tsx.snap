// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`CashHopper tests should group coins 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 pt-4 w-full max-w-full flex-wrap"
    data-testid="cash-hopper"
  >
    <div
      class="relative w-[3rem] h-[3rem]"
    >
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[0px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        1.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-3px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        1.00
      </div>
    </div>
    <div
      class="relative w-[3rem] h-[3rem]"
    >
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[0px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-3px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CashHopper tests should not stack more than 5 coins 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 pt-4 w-full max-w-full flex-wrap"
    data-testid="cash-hopper"
  >
    <div
      class="relative w-[3rem] h-[3rem]"
    >
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[0px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        1.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-3px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        1.00
      </div>
    </div>
    <div
      class="relative w-[3rem] h-[3rem]"
    >
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[0px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-3px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-6px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-9px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="flex items-center bg-gray-500 border-black text-white rounded-full justify-center border w-[3rem] h-[3rem] absolute left-0 top-[-12px]"
        style="transform: perspective(500px) rotateX(50deg);"
      >
        2.00
      </div>
      <div
        class="absolute bottom-[0.5rem] right-[-1px] rounded-full bg-white border border-black text-center w-4 h-4 text-xs"
      >
        7
      </div>
    </div>
  </div>
</DocumentFragment>
`;

exports[`CashHopper tests should render 1`] = `
<DocumentFragment>
  <div
    class="flex gap-2 pt-4 w-full max-w-full flex-wrap"
    data-testid="cash-hopper"
  />
</DocumentFragment>
`;
